### ISSUE TITLE
Training time

### ISSUE THREAD

**Issue by slerman12** [rosinality/relation-networks-pytorch#2]
I'm trying to run this on 4 GPUs, but the training time is intractable (20 days). How many GPUs are required to run this? Or did I make some mistake with configurations? Thanks for any help you can give!

**Comment by rosinality** [comment]
I don't remember the number of GPUs I have used, but this model requires a lot of time. You can may reduce the number of epochs to 300. https://github.com/rosinality/relation-networks-pytorch/blob/master/train.py#L23

**Comment by slerman12** [comment]
Thank you for the speedy response! Do you know if 300 epochs would significantly hurt the performance?

Should I consider other consolidations like reducing the batch size or resizing the images?

**Comment by slerman12** [comment]
One more question (sorry, I'm in a hurry because of an upcoming deadline), do you know how hard it would be to implement distributed training into this? I have access to more GPUs, but they're on different machines. I have no experience with distributed training.

**Comment by rosinality** [comment]
As you can see in the accuracy plot, accuracy improvements after 300 epochs will be not very large. And large batch sizes seems crucial for this model to get a decent performance.

I think implementing distributed training will be not very hard.

**Comment by slerman12** [comment]
Okay, thanks, that's a relief.

**Comment by slerman12** [comment]
Actually, I have one more question. Do you know how much the training time decreases with additional GPUs/nodes? I'm afraid that even using 16 GPUs across 4 nodes would still be too slow. I need to finish training ideally before Friday.

**Comment by rosinality** [comment]
I don't know, but if you use DistributedDataParallel then it will be efficient than current DataParallel based implementation.

**Comment by slerman12** [comment]
I observed that the bottleneck isn't with the GPUs. Interestingly, they're not even in use for long segments of the training. The training seems to pause for awhile, then the GPUs become active for a bit as the training progresses, and then it pauses again, and so on. I looked at the stack trace and it's always at the following line of code where it gets stuck:

`for i, (image, question, q_len, answer, _) in enumerate(pbar):`

Is it possible that it’s the `collate_data` function that caused the problem?

**Comment by slerman12** [comment]
It might be the `sort`.

**Comment by rosinality** [comment]
It could be problematic. But also `__getitem__` of the dataset itself can cause slowdowns. (reading images.) Anyway using distributed training could relieve that issues.

**Comment by slerman12** [comment]
I have 40 CPUs as workers. Shouldn't `__getitem__` be faster in that case? The additional workers don't seem to speed up computation much for some reason.

**Comment by slerman12** [comment]
Here is the stack trace. Do you know where it's getting stuck?

`File "train.py", line 188, in <module>
    train(epoch)
  File "train.py", line 75, in train
    for image, question, q_len, answer, _ in pbar:
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/site-packages/tqdm/std.py", line 1129, in __iter__
    for obj in iterable:
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 345, in __next__
    data = self._next_data()
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 841, in _next_data
    idx, data = self._get_data()
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 808, in _get_data
    success, data = self._try_get_data()
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 761, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/multiprocessing/queues.py", line 107, in get
    if not self._poll(timeout):
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/multiprocessing/connection.py", line 257, in poll
    return self._poll(timeout)
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/multiprocessing/connection.py", line 424, in _poll
    r = wait([self], timeout)
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/multiprocessing/connection.py", line 930, in wait
    ready = selector.select(timeout)
  File "/u/slerman/miniconda3/envs/torch1.5/lib/python3.8/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
KeyboardInterrupt
`

**Comment by rosinality** [comment]
No, increasing num_workers generally can't solve the DataLoader bottlenecks. Maybe you can convert png images to numpy arrays and save it to lmdb, then you can load images from it. Or you can load entire images into the ram if you have enough capacity.

Anyway, distributed training can certainly reduce the data loading times.

**Comment by slerman12** [comment]
I guess one more thing, I get the following warning throughout training: 

```
UserWarning: RNN module weights are not part of single contiguous chunk of memory.
This means they need to be compacted at every call, possibly greatly increasing memory
```

**Comment by rosinality** [comment]
Generally it will not much affect training time.

**Comment by slerman12** [comment]
Hi again, just in case, do you know if training is faster with the implementation you took inspiration from? https://github.com/mesnico/RelationNetworks-CLEVR

**Comment by rosinality** [comment]
I haven't tried that implementation, but I think training speed will be similar.

**Comment by slerman12** [comment]
Hmm, I'm at 126 epochs and accuracy is still at 38%. I'm worried that 300 epochs won't be enough. How consistent are your results?

**Comment by rosinality** [comment]
I think I have only tried 2 times, but these 2 results were similar.

**Comment by slerman12** [comment]
The only change I made is that I tried `non_blocking=True` and `pin_memory=True`. Could that have affected the results? Sorry if this a naive question, I'm new to using GPUs.

**Comment by rosinality** [comment]
Hmm I don't think it could cause much problems.

**Comment by slerman12** [comment]
I'm at 237 epochs and still at 20% accuracy.  I can't think of anything else I might have changed.

**Comment by rosinality** [comment]
Yes, the model should show about 39+% accuracies after 10 epochs. But I don't know why pin_memory or non_blocking could cause the problem.
